### PR TITLE
Eliza Popa/SM-TM-27651/As a Client/Partner I want to have company and country fields on registration

### PR DESCRIPTION
**Pull request by ElizaPopa**

### CODE REVIEW

File: src/main/java/models/AccountDetails.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.WebElement;
+
+import java.util.Random;
+
+public class AccountDetails {
+
+    private String username;
+    private String password;
+    public String countryName;
```
Comment (95teo) [inline src/main/java/models/AccountDetails.java:14]: Why is this public? It should be private too and be accessed via setter/getter only. Thanks!

File: src/main/java/pages/ClientRegistrationPage.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        elementControl.setText(companyRegistrationNumberField, accountDetails.getRegistrationNumber());
+        elementControl.setText(companyTaxationNumberField, accountDetails.getTaxationNumber());
+        elementControl.setText(companyAdressField, accountDetails.getCompanyAddress());
+        elementControl.setText(companyRepresentativeField, accountDetails.getCompanyRepresentative());
+        elementControl.clickElement(termsAndConditionsCheckBox);
+        elementControl.clickElement(enterTheVirtualWorldButton);
+    }
+
+    public void getFreeTrialSubscription(AccountDetails accountDetails) {
+
```
Comment (95teo) [inline src/main/java/pages/ClientRegistrationPage.java:105]: Please remove this unnecessary blank line. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    String errorMessageMinLengthName = ("The min length of the name is 4");
+    String errorMessageMinLengthCompany = "The min lenght of company representative must be 4";
+
+    @Test
+    public void createAClientWithRandomCredentials() throws InterruptedException {
+
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        clientRegistrationPage.getFreeTrialSubscription(accountDetails);
+        Assert.assertTrue(urlSubs.equals("https://dev.xreteh.com/signup"), "The client wasn't properly redirected to the subscription's page");
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:25]: This assertion will always be true. Please use getCurrentUrl() instead. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+public class Tests extends Base {
+
+    String urlSubs = "https://dev.xreteh.com/signup";
+    String errorMessageMinLengthName = ("The min length of the name is 4");
+    String errorMessageMinLengthCompany = "The min lenght of company representative must be 4";
+
+    @Test
+    public void createAClientWithRandomCredentials() throws InterruptedException {
+
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:21]: Please remove this unnecessary blank line. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        accountDetails.setCompanyAddress("");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        String errorMessage = "Company address is required";
+        Assert.assertTrue(errorMessage.equalsIgnoreCase(errorMessage), "Success registration with empty field");
+    }
+
+    @Test
+    public void usingOtherCountryThanRomania() throws InterruptedException {
+
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:40]: Please remove this unnecessary blank line. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        accountDetails.setCountryName("Palau");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        clientRegistrationPage.getFreeTrialSubscription(accountDetails);
+        Assert.assertTrue(urlSubs.equals("https://dev.xreteh.com/signup"), "The client wasn't properly redirected to the subscription's page");
+
+    }
+
+    @Test
+    public void wrongTaxAndRegistrationNumbers() throws InterruptedException {
+
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:51]: Please remove this unnecessary blank line. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void allBoxesRequired() throws InterruptedException {
+
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        accountDetails.setCompanyAddress("");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        String errorMessage = "Company address is required";
+        Assert.assertTrue(errorMessage.equalsIgnoreCase(errorMessage), "Success registration with empty field");
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:35]: This assertion will always be true. Please take one of the messages from a WebElement. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void usingOtherCountryThanRomania() throws InterruptedException {
+
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        accountDetails.setCountryName("Palau");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        clientRegistrationPage.getFreeTrialSubscription(accountDetails);
+        Assert.assertTrue(urlSubs.equals("https://dev.xreteh.com/signup"), "The client wasn't properly redirected to the subscription's page");
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:45]: This assertion will always be true. Please use getCurrentUrl() instead. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Test
+    public void wrongTaxAndRegistrationNumbers() throws InterruptedException {
+
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        accountDetails.setRegistrationNumber("12345678910123456789101234567891031");
+        accountDetails.setTaxationNumber("1239458764532673845967685746352");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        clientRegistrationPage.getFreeTrialSubscription(accountDetails);
+        Assert.assertFalse(Boolean.parseBoolean(urlSubs), "The client shouldn't have been created, there is a minimum of 30 characters that was exceeded");
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:57]: What do you want to do here?

File: src/test/java/companyCountryRegistration/Tests.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        Assert.assertEquals(errorMessageMinLengthCompany, errorMessageMinLengthCompany);
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:58]: This assertion will always be true. Please take one of the messages from a WebElement. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void onlyOneCharacterAdded() throws InterruptedException {
+
+        accountDetails = accountDetailsGenerator.generateOnlyOneRandomCharacterForEachField();
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        elementControl.getTextFromElement(clientRegistrationPage.phoneNumberErrorMessage);
+        Assert.assertEquals(errorMessageMinLengthName,errorMessageMinLengthName);
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:67]: This assertion will always be true. Please take one of the messages from a WebElement. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        accountDetails.setTaxationNumber("1239458764532673845967685746352");
+        clientRegistrationPage.createAClientOnXreteh(accountDetails);
+        clientRegistrationPage.getFreeTrialSubscription(accountDetails);
+        Assert.assertFalse(Boolean.parseBoolean(urlSubs), "The client shouldn't have been created, there is a minimum of 30 characters that was exceeded");
+        Assert.assertEquals(errorMessageMinLengthCompany, errorMessageMinLengthCompany);
+    }
+
+    @Test
+    public void onlyOneCharacterAdded() throws InterruptedException {
+
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:63]: Please remove this unnecessary blank line. Thanks!

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,2 @@
+/.idea
```
Comment (95teo) [inline .gitignore:1]: Try to avoid checking in files from target folder. Thanks!

**Review by 95teo — CHANGES_REQUESTED** [review]

File: src/test/java/companyCountryRegistration/Tests.java (hunk 9 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Tests.java:57]: Nici eu nu stiu exact ce am facut acolo, dar am refacut.

File: src/test/java/companyCountryRegistration/Tests.java (hunk 10 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Tests.java:58]: done

File: src/main/java/models/AccountDetails.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    private String password;
+    private String countryName;
+    private String name;
+    private String phoneNumber;
+    private String companyName;
+    private String registrationNumber;
+    private String taxationNumber;
+    private String companyAddress;
+    private String companyRepresentative;
+    private String emailAddress;
```
Comment (95teo) [inline src/main/java/models/AccountDetails.java:13]: Can you please leave a line between fields and methods? Thanks!

File: src/main/java/modelsGenerator/AccountDetailsGenerator.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
+package modelsGenerator;
```
Comment (95teo) [inline src/main/java/modelsGenerator/AccountDetailsGenerator.java:1]: Can you please reformat this file? Thanks!

File: src/main/java/modelsGenerator/AccountDetailsGenerator.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.github.javafaker.Faker;
+import models.AccountDetails;
+import utils.RandomUtils;
+
+public class AccountDetailsGenerator {
+    AccountDetails accountDetails;
```
Comment (95teo) [inline src/main/java/modelsGenerator/AccountDetailsGenerator.java:8]: Can you please leave a line between fields and methods? Thanks!

File: src/main/java/pages/ClientRegistrationPage.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
+package pages;
```
Comment (95teo) [inline src/main/java/pages/ClientRegistrationPage.java:1]: Can you please reformat this file? Thanks!

File: src/main/java/pages/ClientRegistrationPage.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        elementControl.setText(phoneField, accountDetails.getPhoneNumber());
+        elementControl.setText(companyNameField, accountDetails.getCompanyName());
+        elementControl.setText(companyRegistrationNumberField, accountDetails.getRegistrationNumber());
+        elementControl.setText(companyTaxationNumberField, accountDetails.getTaxationNumber());
+        elementControl.setText(companyAdressField, accountDetails.getCompanyAddress());
+        elementControl.setText(companyRepresentativeField, accountDetails.getCompanyRepresentative());
+        elementControl.clickElement(termsAndConditionsCheckBox);
+        elementControl.clickElement(enterTheVirtualWorldButton);
+    }
+    public void getFreeTrialSubscription(AccountDetails accountDetails) {
```
Comment (95teo) [inline src/main/java/pages/ClientRegistrationPage.java:74]: Please rename this method to clickOnFreeTrialButton(). Also, there is no need to put an AccountDetails parameter. Leave it parameter-less. Thanks!

File: src/test/java/companyCountryRegistration/Base.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    String currentWorkingDirectory;
+    Properties configProperty;
+    String reportFileName;
+    ReportUtils reportUtils;
+    ScreenshotUtils screenshot;
+    ElementControl elementControl;
+    File dir;
+    ClientRegistrationPage clientRegistrationPage;
+    Logger logger;
+    AccountDetailsGenerator accountDetailsGenerator;
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Base.java:38]: I'll put AccountDetailsGenerator and AccountDetails only in the test classes where they are used. We're going to have a common Base for all test classes and these AccountDetailsGenerator and AccountDetails won't be used everywhere. Thanks!

File: src/main/java/pages/ClientRegistrationPage.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    @FindBy(xpath = "//p[normalize-space()='The min length of the name is 4']")
+    private WebElement errorMessageforMinimumLengthForName;
+    @FindBy(xpath = "//p[normalize-space()='The min lenght of company representative must be 4']")
+    private WebElement errorMessageforMinimumLengthForCompanyRepresentative;
+    @FindBy(xpath = "//p[normalize-space()='Please use a valid phone number']")
+    private WebElement phoneNumberErrorMessage;
+    public ClientRegistrationPage(WebDriver driver) {
+        super(driver);
+    }
+    public WebElement getErrorMessageForCompanyAdress() { return errorMessageForCompanyAdress; }
```
Comment (95teo) [inline src/main/java/pages/ClientRegistrationPage.java:51]: Please make these methods to return the text directly from here. No need to do that in the test case class. Thanks!
Applicable for all methods used to get some text/errorMessage

File: src/test/java/companyCountryRegistration/Tests.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
+package companyCountryRegistration;
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:1]: Have you reformated this file?

**Review by 95teo — CHANGES_REQUESTED** [review]

File: src/main/java/pages/ClientRegistrationPage.java (hunk 17 above)
Comment (ElizaPopa) [reply src/main/java/pages/ClientRegistrationPage.java:1]: I've done this and it puts automatically one line break between methods. I hope i don't need to delete them again. :)

File: src/main/java/pages/ClientRegistrationPage.java (hunk 20 above)
Comment (ElizaPopa) [reply src/main/java/pages/ClientRegistrationPage.java:51]: done. Thanks for the hint! It looks better this way.

File: src/test/java/companyCountryRegistration/Tests.java (hunk 21 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Tests.java:1]: I think so… do you refer from Code/ Reformat code, right?

File: src/test/java/companyCountryRegistration/Base.java (hunk 19 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Base.java:38]: Roger that and solved. Thanks!

File: src/test/java/companyCountryRegistration/Tests.java (hunk 21 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Tests.java:1]: I've done also reformat line from Code Menu Button.

File: src/test/java/companyCountryRegistration/Tests.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+    String urlSubs = "https://dev.xreteh.com/signup#";
+    String errorMessageMinLengthName = "The min length of the name is 4";
+    String errorMessageCompanyAdress = "Company address is required";
+    String errorMessageMinLengthCompany = "The min lenght of company representative must be 4";
+    String errorPhoneNumber = "Please use a valid phone number";
+
+    @Test
+    public void createAClientWithRandomCredentials() throws InterruptedException {
+        accountDetails = accountDetailsGenerator.generateRandomAccountDetails();
+        accountDetailsGenerator = new AccountDetailsGenerator();
```
Comment (95teo) [inline src/test/java/companyCountryRegistration/Tests.java:21]: This line should be above the previous one. Please check for all methods on this test class. Thanks!

**Review by 95teo — CHANGES_REQUESTED** [review]

File: src/test/java/companyCountryRegistration/Tests.java (hunk 22 above)
Comment (ElizaPopa) [reply src/test/java/companyCountryRegistration/Tests.java:21]: done

**Review by 95teo — APPROVED** [review]

**Review by pistamarcel — APPROVED** [review]

**Pull request merged**